### PR TITLE
LLK implementation for EQ, NE - integer datatypes

### PR DESCRIPTION
**Pull request by KalaivaniMCW**
### PR Category

     
Bug fix

### Summary

    
 Issue https://github.com/tenstorrent/tt-metal/issues/23179 #23333 

### Notes for reviewers

This PR replaces the previous sub(A,B) + eqz/nez implementation of ttnn.eq / ttnn.ne for integer dtypes (INT32, UINT32, UINT16) with a native SFPU kernel. 
- The old path could produce incorrect results because the subtraction was computed through a float-capable path that loses precision/overflows for large integers; 
- the new `calculate_binary_eq_int` performs a bitwise XOR followed by a conditional store, comparing the full integer bit pattern correctly. 

Changes:

- Route integer EQ/NE through SFPU (eq_int_tile/ne_int_tile) instead of the FPU SUB + EQZ/NEZ postprocess, and add the eq_int/ne_int SfpuTypes plus addrmod configuration on both archs.
- Add the new XOR-based calculate_binary_eq_int SFPU kernel and the eq_int_tile/ne_int_tile(_init) tile APIs.
- Extend the UInt16 scalar-broadcast software-fallback condition to cover EQ/NE, and update int32 relational/arithmetic tests with overflow boundary values (e.g. 2147483647, -2147483648).

**Perf: Wormhole** 

**NE:**
Input Shape | main (ns) | SFPU (ns) | Saved (ns) | Improvement
-- | -- | -- | -- | --
(32, 32) | 3,239 | 3,110 | 129 | 3.98%
(128, 128) | 3,684 | 3,487 | 197 | 5.35%
(1024, 1024) | 79,571 | 78,919 | 652 | 0.82%

Adding replay buffer: 
image

**EQ:**

Input Shape | main (ns) | SFPU (ns) | Saved (ns) | Improvement
-- | -- | -- | -- | --
(32, 32) | 3,214 | 2,935 | 279 | 8.68%
(128, 128) | 3,887 | 3,583 | 304 | 7.82%
(1024, 1024) | 80,700 | 79,161 | 1,539 | 1.91%

image

**Perf (BH):**
image

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:kalaivani/eq_ne_sfpu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:kalaivani/eq_ne_sfpu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:kalaivani/eq_ne_sfpu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:kalaivani/eq_ne_sfpu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:kalaivani/eq_ne_sfpu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:kalaivani/eq_ne_sfpu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:kalaivani/eq_ne_sfpu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:kalaivani/eq_ne_sfpu)

### CODE REVIEW

**Review by Aswinmcw — APPROVED** [review]

**Review by VirdhatchaniKN — APPROVED** [review]

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h [hunk 1]
```diff
@@ -185,6 +185,38 @@ inline void calculate_binary_comp_fp32(const uint dst_index_in0, const uint dst_
     }
 }
 
+// Int32 equality comparisons: eq(a,b) and ne(a,b).
+// XOR a and b; the result is zero iff a == b. A conditional store writes
```
Comment (KalaivaniMCW) [inline tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h]: if*

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h [hunk 2]
```diff
@@ -181,6 +181,38 @@ inline void calculate_binary_comp_fp32(const uint dst_index_in0, const uint dst_
     }
 }
 
+// Int32 equality comparisons: eq(a,b) and ne(a,b).
+// XOR a and b; the result is zero iff a == b. A conditional store writes
```
Comment (KalaivaniMCW) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h]: if*

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h [hunk 3]
```diff
@@ -181,6 +181,38 @@ inline void calculate_binary_comp_fp32(const uint dst_index_in0, const uint dst_
...
+    constexpr uint b = p_sfpu::LREG1;
+    constexpr uint one = p_sfpu::LREG2;
+    constexpr uint dst_tile_size = 64;
+
+    constexpr bool is_eq = (RELATIONAL_OP == SfpuType::eq);
+    constexpr uint default_result = is_eq ? p_sfpu::LCONST_0 : one;
+    constexpr uint equal_result = is_eq ? one : p_sfpu::LCONST_0;
+
+    TTI_SFPLOADI(one, sfpi::SFPLOADI_MOD0_UPPER, 0x0000);
+    TTI_SFPLOADI(one, sfpi::SFPLOADI_MOD0_LOWER, 0x0001);
```
Comment (KalaivaniMCW) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h]: Any better ways to load 1 into LREG2 ?

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h [hunk 4]
```diff
@@ -280,4 +312,41 @@ inline void calculate_binary_comp_uint(const uint dst_index_in0, const uint dst_
         }
     }
 }
+
+// UInt32/UInt16 equality comparisons: eq(a,b) and ne(a,b).
+// XOR a and b; the result is zero iff a == b. A conditional store writes
```
Comment (KalaivaniMCW) [inline tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h]: if*

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h [hunk 5]
```diff
@@ -276,4 +308,41 @@ inline void calculate_binary_comp_uint(const uint dst_index_in0, const uint dst_
         }
     }
 }
+
+// UInt32/UInt16 equality comparisons: eq(a,b) and ne(a,b).
+// XOR a and b; the result is zero iff a == b. A conditional store writes
```
Comment (KalaivaniMCW) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h]: if*

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h (hunk 3 above)
Comment (Aswinmcw) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h]: Single instruction alternative:

```cpp
TTI_SFPLOADI(one, sfpi::SFPLOADI_MOD0_USHORT, 0x0001);
```

`SFPLOADI_MOD0_USHORT` (mode 2) does `LReg[VD].u32 = ZeroExtend(Imm16)` per the ISA — zero-extends the 16-bit immediate to a full 32-bit unsigned integer. Replaces the two-instruction UPPER+LOWER sequence.

Already used in the codebase for the same purpose — e.g. Quasar `ckernel_sfpu_binary_comp.h:36`:
```cpp
TTI_SFPLOADI(p_sfpu::LREG7, sfpi::SFPLOADI_MOD0_USHORT, 0x01);
```

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h [hunk 6]
```diff
@@ -276,4 +276,40 @@ inline void calculate_binary_comp_uint(const uint dst_index_in0, const uint dst_
...
+
+    constexpr bool is_eq = (RELATIONAL_OP == SfpuType::eq);
+    constexpr uint default_result = is_eq ? p_sfpu::LCONST_0 : one;
+    constexpr uint equal_result = is_eq ? one : p_sfpu::LCONST_0;
+
+    TTI_SFPLOADI(one, sfpi::SFPLOADI_MOD0_USHORT, 0x0001);
+
+#pragma GCC unroll 8
+    for (int d = 0; d < ITERATIONS; d++) {
+        TT_SFPLOAD(a, ld_st_mod, ADDR_MOD_7, dst_index_in0 * dst_tile_size);
```
Comment (fvranicTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h:305]: Have you considered using the replay buffer? That should improve performance of this kernel. SFPI automatically does this, but not when the code is written in TTI intrinsics.

Comment (fvranicTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h:305]: Something like:

```cpp
constexpr uint binary_eq_int_replay_len = 7;

template <SfpuType RELATIONAL_OP, DataFormat DATA_FORMAT>
inline void calculate_binary_eq_int_init(
    const uint dst_index_in0, const uint dst_index_in1, const uint dst_index_out) {
    static_assert(
        DATA_FORMAT == DataFormat::Int32 || DATA_FORMAT == DataFormat::UInt16 || DATA_FORMAT == DataFormat::UInt32,
        "Unsupported data format for calculate_binary_eq_int_init(). Supported formats: Int32, UInt16, UInt32.");
    static_assert(RELATIONAL_OP == SfpuType::eq || RELATIONAL_OP == SfpuType::ne, "Supported operation types: eq, ne");

    constexpr InstrModLoadStore ld_st_mod =
        (DATA_FORMAT == DataFormat::UInt16) ? InstrModLoadStore::LO16 : InstrModLoadStore::INT32;
    constexpr uint a = p_sfpu::LREG0;
    constexpr uint b = p_sfpu::LREG1;
    constexpr uint one = p_sfpu::LREG2;
    constexpr uint dst_tile_size = 64;

    constexpr bool is_eq = (RELATIONAL_OP == SfpuType::eq);
    constexpr uint default_result = is_eq ? p_sfpu::LCONST_0 : one;
    constexpr uint equal_result = is_eq ? one : p_sfpu::LCONST_0;

    const uint off_in0 = dst_index_in0 * dst_tile_size;
    const uint off_in1 = dst_index_in1 * dst_tile_size;
    const uint off_out = dst_index_out * dst_tile_size;

    load_replay_buf(0, binary_eq_int_replay_len, [off_in0, off_in1, off_out] {
        TT_SFPLOAD(a, ld_st_mod, ADDR_MOD_7, off_in0);
        TT_SFPLOAD(b, ld_st_mod, ADDR_MOD_7, off_in1);
        TT_SFPSTORE(default_result, ld_st_mod, ADDR_MOD_7, off_out);
        TTI_SFPXOR(0, b, a, 0);
        TTI_SFPSETCC(0, a, 0, sfpi::SFPSETCC_MOD1_LREG_EQ0);
        TT_SFPSTORE(equal_result, ld_st_mod, ADDR_MOD_6, off_out);
        TTI_SFPENCC(0, 0, 0, 0);
    });
}

template <bool APPROXIMATION_MODE, int ITERATIONS, SfpuType RELATIONAL_OP, DataFormat DATA_FORMAT>
inline void calculate_binary_eq_int(
    const uint dst_index_in0, const uint dst_index_in1, const uint dst_index_out) {
    static_assert(
        DATA_FORMAT == DataFormat::Int32 || DATA_FORMAT == DataFormat::UInt16 || DATA_FORMAT == DataFormat::UInt32,
        "Unsupported data format for calculate_binary_eq_int(). Supported formats: Int32, UInt16, UInt32.");
    static_assert(RELATIONAL_OP == SfpuType::eq || RELATIONAL_OP == SfpuType::ne, "Supported operation types: eq, ne");

    TTI_SFPLOADI(p_sfpu::LREG2, sfpi::SFPLOADI_MOD0_USHORT, 0x0001);

#pragma GCC unroll 8
    for (int d = 0; d < ITERATIONS; d++) {
        lltt::replay(0, binary_eq_int_replay_len);
    }
}
```

Comment (KalaivaniMCW) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h:305]: Sure, I'll try this and test

Comment (fvranicTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h:305]: Hm can’t we have an init method that sets the buffer once instead on every call?

**Review by mcw-anasuya — APPROVED** [review]

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h (hunk 6 above)
Comment (KalaivaniMCW) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h:305]: I haven't tried modifying the inits. Since we need the input and output dst tile offsets, offsets idst0/idst1/odst, are currently passed as runtime arguments, and not passed to the init. 

But based on using the replay buffer, I didn't see any improvement in BH, the TTI intrinsics is slightly faster. 
In WH, there is a slight difference at times. 

Did you want me to try with `SFPLOADMACRO` ? If so, I can try that  using `ckernel_sfpu_binary_max_min.h` as reference in a follow-up.

Comment (fvranicTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h:305]: Ok, Thanks for trying this out. Please don't use loadmacro.

Comment (fvranicTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h:305]: @KalaivaniMCW if I am interpreting the results Well, implementation with replay buffer is slower. Let's use the original approach instead.

Comment (KalaivaniMCW) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h:305]: yes @fvranicTT  I dropped the replay buffer changes

**Review by fvranicTT — APPROVED** [review]

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h [hunk 7]
```diff
@@ -280,4 +280,40 @@ inline void calculate_binary_comp_uint(const uint dst_index_in0, const uint dst_
...
+    constexpr uint a = p_sfpu::LREG0;
+    constexpr uint b = p_sfpu::LREG1;
+    constexpr uint one = p_sfpu::LREG2;
+    constexpr uint dst_tile_size = 64;
+
+    constexpr bool is_eq = (RELATIONAL_OP == SfpuType::eq);
+    constexpr uint default_result = is_eq ? p_sfpu::LCONST_0 : one;
+    constexpr uint equal_result = is_eq ? one : p_sfpu::LCONST_0;
+
+    TTI_SFPLOADI(one, sfpi::SFPLOADI_MOD0_USHORT, 0x0001);
```
Comment (amahmudTT) [inline tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_comp.h:305]: Why not have this in the init function, (same comment if such instances exist)
